PyPy test failures: cannot overwrite an attribute

Testing with PyPy 3.7 v7.3.1 on lnux 64, I have a number of failures. xref gh-1012 for a complete report of an older version on windows. Here I would like to focus on the following failure (from test_attrs.py test_overwrite):
```
from h5py import File
f = File('tempname', 'w')
f.attrs['a'] = 4.0
f.attrs['a'] = 5.0
assert f.attrs['a'] == 5.0
```
The old value is preserved. It seems this is connected to `h5a.delete` somehow? Any hints what could be the problem? Is there a flush or other operation that is dependent on refcount semantics when an object goes out of scope?

Thanks. Before the PR there are 9 failures, and with it there are 9 failed, 1 error:
```
================== short test summary info ====================
FAILED h5py/tests/test_attrs.py::TestAccess::test_modify - AssertionError: 3 != 4
FAILED h5py/tests/test_attrs.py::TestCreate::test_empty - AssertionError: Empty(dtype=dtype('<f4')) != Empty(dtype=dtype('float6...
FAILED h5py/tests/test_attrs_data.py::TestTypes::test_complex - AssertionError: dtype('complex64') != dtype('>c8')
FAILED h5py/tests/test_attrs_data.py::TestTypes::test_float - AssertionError: dtype('<f4') != dtype('>f4')
FAILED h5py/tests/test_attrs_data.py::TestTypes::test_int - AssertionError: dtype('int8') != <class 'numpy.int16'>
FAILED h5py/tests/test_attrs_data.py::TestTypes::test_string - AssertionError: dtype('S1') != dtype('S10')
FAILED h5py/tests/test_file2.py::TestDealloc::test_autoclose - AssertionError: 3 != 2
FAILED h5py/tests/test_file2.py::TestFileObj::test_TemporaryFile - AssertionError: True is not false
FAILED h5py/tests/test_vds/test_lowlevel_vds.py::test_virtual_prefix - AssertionError: 
ERROR h5py/tests/test_attrs_data.py::TestTypes::test_int - RuntimeError: Unspecified error in H5Iget_type (return value <0)
```
It seems closing the attr is not sufficient?

--- a/h5py/_hl/attrs.py
+++ b/h5py/_hl/attrs.py
@@ -215,6 +215,8 @@ class AttributeManager(base.MutableMappingHDF5, base.CommonStateObject):
                         attr.close()
                         h5a.delete(self._id, self._e(tempname))
                         raise
+                finally:
+                    attr.close()
 
     def modify(self, name, value):
         """ Change the value of an attribute while preserving its type.

--- a/h5py/_hl/attrs.py
+++ b/h5py/_hl/attrs.py
@@ -193,30 +193,26 @@ class AttributeManager(base.MutableMappingHDF5, base.CommonStateObject):
 
             tempname = uuid.uuid4().hex
 
+            attr = h5a.create(self._id, self._e(tempname), htype, space)
             try:
-                attr = h5a.create(self._id, self._e(tempname), htype, space)
+                if not isinstance(data, Empty):
+                    attr.write(data, mtype=htype2)
             except:
+                attr.close()
+                h5a.delete(self._id, self._e(tempname))
                 raise
             else:
                 try:
-                    if not isinstance(data, Empty):
-                        attr.write(data, mtype=htype2)
+                    # No atomic rename in HDF5 :(
+                    if h5a.exists(self._id, self._e(name)):
+                        h5a.delete(self._id, self._e(name))
+                    h5a.rename(self._id, self._e(tempname), self._e(name))
                 except:
                     attr.close()
                     h5a.delete(self._id, self._e(tempname))
                     raise
-                else:
-                    try:
-                        # No atomic rename in HDF5 :(
-                        if h5a.exists(self._id, self._e(name)):
-                            h5a.delete(self._id, self._e(name))
-                        h5a.rename(self._id, self._e(tempname), self._e(name))
-                    except:
-                        attr.close()
-                        h5a.delete(self._id, self._e(tempname))
-                        raise
-                finally:
-                    attr.close()
+            finally:
+                attr.close()
 
     def modify(self, name, value):
         """ Change the value of an attribute while preserving its type.

